sstables/mp_row_consumer.cc

```diff
@@ -80,9 +81,35 @@ atomic_cell make_counter_cell(api::timestamp_type timestamp, bytes_view value) {
     return ccb.build(timestamp);
 }
 
+// See #6130.
+static data_type freeze_types_in_collections(data_type t) {
+    return ::visit(*t, make_visitor(
+    [] (const map_type_impl& typ) -> data_type {
+        return map_type_impl::get_instance(
+                freeze_types_in_collections(typ.get_keys_type()->freeze()),
+                freeze_types_in_collections(typ.get_values_type()->freeze()),
+                typ.is_multi_cell());
```

Comment:
Does the bug affect only non-frozen collections or frozen as well? If it affects only non-frozen collections then maybe we could break out of the recursion earlier.

Reply:
affects everything